67_SubAllocatedDescriptorSet/main.cpp

```diff
@@ -0,0 +1,190 @@
+		constexpr static inline uint32_t MinDescriptorSetAllocationSize = 1u;
+
+	public:
+		// Yay thanks to multiple inheritance we cannot forward ctors anymore
+		SubAllocatedDescriptorSetApp(const path& _localInputCWD, const path& _localOutputCWD, const path& _sharedInputCWD, const path& _sharedOutputCWD) :
+			system::IApplicationFramework(_localInputCWD,_localOutputCWD,_sharedInputCWD,_sharedOutputCWD) {}
+
+		// we stuff all our work here because its a "single shot" app
+		bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+		{
+			using nbl::video::IGPUDescriptorSetLayout;
+
+			// Remember to call the base class initialization!
+			if (!device_base_t::onAppInitialized(std::move(system)))
+				return false;
+			if (!asset_base_t::onAppInitialized(std::move(system)))
+				return false;
+
+
+			// We'll allow subsequent iterations to overlap each other on the GPU, the only limiting factors are
+			// the amount of memory in the streaming buffers and the number of commandpools we can use simultaenously.
+			constexpr auto MaxConcurrency = 64;
+
+			// Since this time we don't throw the Command Pools away and we'll reset them instead, we don't create the pools with the transient flag
+			m_poolCache = ICommandPoolCache::create(core::smart_refctd_ptr(m_device),getComputeQueue()->getFamilyIndex(),IGPUCommandPool::CREATE_FLAGS::NONE,MaxConcurrency);
+
+			// In contrast to fences, we just need one semaphore to rule all dispatches
+			m_timeline = m_device->createSemaphore(m_iteration);
+
+			// Descriptor set sub allocator
+
+			video::IGPUDescriptorSetLayout::SBinding bindings[1];
+			{
+				bindings[0].binding = 0;
+				bindings[0].count = 65536u;
+				bindings[0].createFlags = core::bitflag(IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_AFTER_BIND_BIT) 
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_UNUSED_WHILE_PENDING_BIT 
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_PARTIALLY_BOUND_BIT;
+				bindings[0].type = asset::IDescriptor::E_TYPE::ET_STORAGE_IMAGE;
+				bindings[0].stageFlags = asset::IShader::E_SHADER_STAGE::ESS_COMPUTE;
+			}
+
+			std::span<video::IGPUDescriptorSetLayout::SBinding> bindingsSpan(bindings);
+
+			// TODO: I don't think these are needed for sub allocated descriptor sets (alignment isn't needed, and min size is 1)
+			auto subAllocatedDescriptorSet = core::make_smart_refctd_ptr<nbl::video::SubAllocatedDescriptorSet>(
+				bindings, MaxDescriptorSetAllocationAlignment, MinDescriptorSetAllocationSize
+			);
+
+			std::vector<uint32_t> allocation, size;
+			{
+				for (uint32_t i = 0; i < 512; i++)
+				{
+					allocation.push_back(core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address);
+					size.push_back(4);
+				}
+				subAllocatedDescriptorSet->multi_allocate(0, allocation.size(), &allocation[0], &size[0]);
```

Comment:
whats the leading 0?
also there should be no reason for the size, but I'll leave that for the main PR

Reply:
> whats the leading 0?
ah ok its the binding you're allocating from, please use a named variable/constant